rfcs/2020-08-27-3603-postgres-metrics.md

```diff
@@ -0,0 +1,157 @@
+# RFC 3603 - 2020-08-27 - Collecting metrics from PostgreSQL
+
+This RFC is to introduce a new metrics source to consume metrics from PostgreSQL database servers. The high level plan is to implement one source that collects metrics from PostgreSQL server instances.
+
+Background reading on PostgreSQL monitoring:
+
+- https://www.datadoghq.com/blog/postgresql-monitoring/
+
+## Scope
+
+This RFC will cover:
+
+- A new source for PostgreSQL server metrics.
+
+This RFC will not cover:
+
+- Other databases.
+
+## Motivation
+
+Users want to collect, transform, and forward metrics to better observe how their PostgreSQL databases are performing.
+
+## Internal Proposal
+
+Build a single source called `postgresql_metrics` (name to be confirmed) to collect PostgreSQL metrics. We support all non-EOL'ed PostgreSQL versions.
+
+The recommended implementation is to use the Rust PostgreSQL client to connect the target database server by address specified in configuration.
+
+- https://docs.rs/postgres/0.17.5/postgres/index.html
+
+
+The source would then run the following queries:
+
+- `SELECT * FROM pg_stat_database`
+- `SELECT * FROM pg_stat_database_conflicts`
+- `SELECT * FROM pg_stat_bgwriter`
+
+And return these metrics by parsing the query results and converting them into metrics using the database name and column names.
+
+- `pg_up` -> Used as an uptime metric with 0 for successful collection and 1 for failed collection (gauge)
+- `pg_stat_database_blk_read_time_seconds_total` tagged with db, host, server, user (counter)
+- `pg_stat_database_blk_write_time_seconds_total` tagged with db, host, server, user (counter)
+- `pg_stat_database_blks_hit_total` tagged with db, host, server, user (counter)
+- `pg_stat_database_blks_read_total` tagged with db, host, server, user (counter)
+- `pg_stat_database_stats_reset` tagged with db, host, server, user(counter)
+- `pg_stat_bgwriter_buffers_alloc_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_buffers_backend_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_buffers_backend_fsync_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_buffers_checkpoint_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_buffers_clean_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_checkpoint_sync_time_seconds_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_checkpoint_write_time_seconds_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_checkpoints_req_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_checkpoints_timed_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_maxwritten_clean_total` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_stats_reset` (counter)
```

Comment:
Should this be `_total`?

Reply:
No - stats_reset is a timestamp technically.

Reply:
Ah, yeah, I see it. I'm not sure how we'd represent that as a counter then. It seems to me like it should be a gauge (which is what it appears the other exporters do).

An alternative would be to make it `pg_stat_bgwriter_stats_reset_seconds_total` which would represent the seconds since the last reset and would be a counter.

Reply:
I've updated it to be a gauge. It feels like folks can manipulate it but we should pass through the original.